src/ch/ch_monitor.c

```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+                if ((ev == virCHMonitorVmmEventShutdown ||
+                     state == VIR_DOMAIN_SHUTDOWN) &&
+                      (virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0)) {
+
+                    virCHProcessStop(driver, vm, VIR_DOMAIN_SHUTOFF_SHUTDOWN);
+
+                    virCHDomainObjEndJob(vm);
+                }
+                virObjectUnlock(vm);
+                break;
+            }
+        case virCHMonitorVmEventBooting:
+        case virCHMonitorVmEventPausing:
+        case virCHMonitorVmEventPaused:
+        case virCHMonitorVmEventResuming:
+        case virCHMonitorVmEventSnapshotting:
+        case virCHMonitorVmEventSnapshotted:
+        case virCHMonitorVmEventRestoring:
+        case virCHMonitorVmEventResizing:
+        case virCHMonitorVmEventResized:
+        case virCHMonitorVmEventDeleted:
+        case virCHMonitorVmmEventStarting:
+        case virCHMonitorCpuCreateVcpu:
+            break;
+        case virCHMonitorEventMax:
+        default:
+            VIR_WARN("unkown event from monitor!");
+            break;
+    }
+
+    return 0;
+}
+
+/*
+ * Helper function to find a block of valid JSON
+ * from a stream of multiple JSON blocks.
+ */
+static inline char *end_of_json(char *str, size_t len)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while ((i < len) && (!started || blocks > 0)) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
+}
+
+/*
+ * Caller should have reference on Monitor and Domain
+ */
+static int virCHMonitorProcessEvents(virCHMonitorPtr mon, int events)
+{
+    virJSONValuePtr obj = NULL;
+    char *buf = mon->buffer;
+    int ret = 0;
+    int i = 0;
+    size_t sz;
+
+    for (i = 0; i < events; i++) {
+        char tmp;
+        char *end = end_of_json(buf, mon->buf_fill_sz);
+
+        /*
+         * end should never be NULL! We validated that there
+         * is a valid JSON document before calling end_of_json,
+         * and end_of_json returns NULL only if it cannot find
+         * a valid JSON document.
+         */
+        assert(end);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnull-dereference"
+        tmp = *end, *end = 0;
+#pragma GCC diagnostic pop
+
+        if ((obj = virJSONValueFromString(buf))) {
+            if (virCHMonitorProcessEvent(mon, obj) < 0) {
+                VIR_WARN("Failed to process the event!");
+                ret = -1;
+            }
+            virJSONValueFree(obj);
+        } else {
+            VIR_WARN("Invalid JSON from monitor");
+            ret = -1;
+        }
+
+        *end = tmp;
+        buf = end;
+    }
+
+    /*
+     * If the buffer still has incomplete data, lets
+     * push it to the beginning.
+     */
+    sz = buf - mon->buffer;
```

Comment:
I think `sz` is the amount of data that was processed (`mon->buffer` is the start and `buf` is the location after the last json event that was processed).